File: packages/go/pkg/bot/chat_management.go
package bot

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/Nam088/telebot/packages/go/pkg/types"
)

// SetChatTitle changes the title of a chat.
// Titles can't be changed for private chats.
// The bot must be an administrator in the chat for this to work.
func (b *Bot) SetChatTitle(ctx context.Context, opts *types.SetChatTitleOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "setChatTitle", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// SetChatDescription changes the description of a group, a supergroup or a channel.
// The bot must be an administrator in the chat for this to work.
func (b *Bot) SetChatDescription(ctx context.Context, opts *types.SetChatDescriptionOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "setChatDescription", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// SetChatPhoto sets a new profile photo for the chat.
// Photos can't be changed for private chats.
// The bot must be an administrator in the chat for this to work.
func (b *Bot) SetChatPhoto(ctx context.Context, opts *types.SetChatPhotoOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "setChatPhoto", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// DeleteChatPhoto deletes a chat photo.
// Photos can't be changed for private chats.
// The bot must be an administrator in the chat for this to work.
func (b *Bot) DeleteChatPhoto(ctx context.Context, opts *types.DeleteChatPhotoOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "deleteChatPhoto", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// PinChatMessage adds a message to the list of pinned messages in a chat.
// If the chat is not a private chat, the bot must be an administrator.
func (b *Bot) PinChatMessage(ctx context.Context, opts *types.PinChatMessageOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "pinChatMessage", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// UnpinChatMessage removes a message from the list of pinned messages in a chat.
// If the chat is not a private chat, the bot must be an administrator.
func (b *Bot) UnpinChatMessage(ctx context.Context, opts *types.UnpinChatMessageOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "unpinChatMessage", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// UnpinAllChatMessages clears the list of pinned messages in a chat.
// If the chat is not a private chat, the bot must be an administrator.
func (b *Bot) UnpinAllChatMessages(ctx context.Context, opts *types.UnpinAllChatMessagesOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "unpinAllChatMessages", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// SetChatPermissions sets default chat permissions for all members.
// The bot must be an administrator in the group or a supergroup.
func (b *Bot) SetChatPermissions(ctx context.Context, opts *types.SetChatPermissionsOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "setChatPermissions", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// ExportChatInviteLink generates a new primary invite link for a chat.
// Any previously generated primary link is revoked.
// The bot must be an administrator in the chat for this to work.
func (b *Bot) ExportChatInviteLink(ctx context.Context, opts *types.ExportChatInviteLinkOptions) (string, error) {
	var link string
	if err := b.Request(ctx, "exportChatInviteLink", opts, &link); err != nil {
		return "", err
	}
	return link, nil
}

// SetChatMenuButton changes the bot's menu button in a private chat, or the default menu button.
func (b *Bot) SetChatMenuButton(ctx context.Context, opts *types.SetChatMenuButtonOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "setChatMenuButton", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// GetChatMenuButton gets the current value of the bot's menu button in a private chat, or the default menu button.
func (b *Bot) GetChatMenuButton(ctx context.Context, opts *types.GetChatMenuButtonOptions) (types.MenuButton, error) {
	var raw json.RawMessage
	if err := b.Request(ctx, "getChatMenuButton", opts, &raw); err != nil {
		return nil, err
	}

	return unmarshalMenuButton(raw)
}

// unmarshalMenuButton converts the raw JSON returned by getChatMenuButton into
// one of the concrete MenuButton implementations based on the "type" field.
func unmarshalMenuButton(raw json.RawMessage) (types.MenuButton, error) {
	var envelope struct {
		Type string `json:"type"`
	}
	if err := json.Unmarshal(raw, &envelope); err != nil {
		return nil, fmt.Errorf("failed to decode menu button type: %w", err)
	}

	switch envelope.Type {
	case "commands":
		var btn types.MenuButtonCommands
		if err := json.Unmarshal(raw, &btn); err != nil {
			return nil, fmt.Errorf("failed to decode commands menu button: %w", err)
		}
		return btn, nil
	case "web_app":
		var btn types.MenuButtonWebApp
		if err := json.Unmarshal(raw, &btn); err != nil {
			return nil, fmt.Errorf("failed to decode web_app menu button: %w", err)
		}
		return btn, nil
	case "default":
		var btn types.MenuButtonDefault
		if err := json.Unmarshal(raw, &btn); err != nil {
			return nil, fmt.Errorf("failed to decode default menu button: %w", err)
		}
		return btn, nil
	default:
		return nil, fmt.Errorf("unsupported menu button type: %s", envelope.Type)
	}
}

// SetMyDefaultAdministratorRights changes the default administrator rights of the bot.
func (b *Bot) SetMyDefaultAdministratorRights(ctx context.Context, opts *types.SetMyDefaultAdministratorRightsOptions) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "setMyDefaultAdministratorRights", opts, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// GetMyDefaultAdministratorRights gets the current default administrator rights of the bot.
func (b *Bot) GetMyDefaultAdministratorRights(ctx context.Context, opts *types.GetMyDefaultAdministratorRightsOptions) (*types.ChatAdministratorRights, error) {
	var rights types.ChatAdministratorRights
	if err := b.Request(ctx, "getMyDefaultAdministratorRights", opts, &rights); err != nil {
		return nil, err
	}
	return &rights, nil
}

// SetChatStickerSet sets a new group sticker set for a supergroup.
// The bot must be an administrator in the chat for this to work and must have
// the can_manage_chat administrator right. Use the field "active_sticker_set_name"
// returned by getChat to check the name of the current group sticker set.
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - chatID: Unique identifier for the target chat.
//   - stickerSetName: Name of the sticker set to be set as the group sticker set.
//
// Returns:
//   - bool: True on success.
//   - error: TelegramError if the API returns an error.
//
// Example:
//
//	ok, err := b.SetChatStickerSet(ctx, int64(-1001234567890), "TelebotTestSet")
func (b *Bot) SetChatStickerSet(ctx context.Context, chatID any, stickerSetName string) (bool, error) {
	payload := map[string]any{
		"chat_id":          chatID,
		"sticker_set_name": stickerSetName,
	}
	var ok bool
	if err := b.Request(ctx, "setChatStickerSet", payload, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// DeleteChatStickerSet deletes a group sticker set from a supergroup.
// The bot must be an administrator in the chat for this to work and must have
// the can_manage_chat administrator right.
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - chatID: Unique identifier for the target chat.
//
// Returns:
//   - bool: True on success.
//   - error: TelegramError if the API returns an error.
//
// Example:
//
//	ok, err := b.DeleteChatStickerSet(ctx, int64(-1001234567890))
func (b *Bot) DeleteChatStickerSet(ctx context.Context, chatID any) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "deleteChatStickerSet", map[string]any{"chat_id": chatID}, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// verifyRequest performs a verifyUser or verifyChat request on behalf of the
// organization that owns the bot, sending the optional custom description only
// when the caller supplied a non-empty one.
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - method: Bot API method name to call.
//   - idKey: Wire key identifying the target, "user_id" or "chat_id".
//   - id: Identifier of the target user or chat.
//   - customDescription: Custom description for the verification status, 0-70
//     characters; omitted when empty.
//
// Returns:
//   - bool: True on success.
//   - error: TelegramError if the API returns an error.
func (b *Bot) verifyRequest(ctx context.Context, method, idKey string, id any, customDescription string) (bool, error) {
	payload := map[string]any{idKey: id}
	if customDescription != "" {
		payload["custom_description"] = customDescription
	}
	var ok bool
	if err := b.Request(ctx, method, payload, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// VerifyUser verifies a user on behalf of the organization which owns the bot.
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - userID: Unique identifier of the target user.
//   - customDescription: Custom description for the verification status, 0-70
//     characters; pass an empty string to omit it.
//
// Returns:
//   - bool: True on success.
//   - error: TelegramError if the API returns an error.
//
// Example:
//
//	ok, err := b.VerifyUser(ctx, 123456, "Official Staff")
func (b *Bot) VerifyUser(ctx context.Context, userID int64, customDescription string) (bool, error) {
	return b.verifyRequest(ctx, "verifyUser", "user_id", userID, customDescription)
}

// VerifyChat verifies a chat on behalf of the organization which owns the bot.
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - chatID: Unique identifier for the target chat or username of the target
//     channel (in the format @channelusername).
//   - customDescription: Custom description for the verification status, 0-70
//     characters; pass an empty string to omit it.
//
// Returns:
//   - bool: True on success.
//   - error: TelegramError if the API returns an error.
//
// Example:
//
//	ok, err := b.VerifyChat(ctx, "@channel", "Verified Community")
func (b *Bot) VerifyChat(ctx context.Context, chatID any, customDescription string) (bool, error) {
	return b.verifyRequest(ctx, "verifyChat", "chat_id", chatID, customDescription)
}

// RemoveUserVerification removes verification from a user that was previously
// verified on behalf of the organization which owns the bot. Does not affect
// independently obtained verification.
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - userID: Unique identifier of the target user.
//
// Returns:
//   - bool: True on success.
//   - error: TelegramError if the API returns an error.
//
// Example:
//
//	ok, err := b.RemoveUserVerification(ctx, 123456)
func (b *Bot) RemoveUserVerification(ctx context.Context, userID int64) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "removeUserVerification", map[string]any{"user_id": userID}, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// RemoveChatVerification removes verification from a chat that was previously
// verified on behalf of the organization which owns the bot. Does not affect
// independently obtained verification.
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - chatID: Unique identifier for the target chat or username of the target
//     channel (in the format @channelusername).
//
// Returns:
//   - bool: True on success.
//   - error: TelegramError if the API returns an error.
//
// Example:
//
//	ok, err := b.RemoveChatVerification(ctx, "@channel")
func (b *Bot) RemoveChatVerification(ctx context.Context, chatID any) (bool, error) {
	var ok bool
	if err := b.Request(ctx, "removeChatVerification", map[string]any{"chat_id": chatID}, &ok); err != nil {
		return false, err
	}
	return ok, nil
}

// GetUserChatBoosts retrieves the list of boosts added to a chat by a user.
// Requires administrator rights in the chat.
//
// Parameters:
//   - ctx: Context for cancellation and timeout.
//   - chatID: Unique identifier for the chat or username of the channel (in the
//     format @channelusername).
//   - userID: Unique identifier of the target user.
//
// Returns:
//   - *types.UserChatBoosts: The list of boosts on success.
//   - error: TelegramError if the API returns an error.
//
// Example:
//
//	boosts, err := b.GetUserChatBoosts(ctx, int64(-1001234567890), 123456)
//	fmt.Println(len(boosts.Boosts))
func (b *Bot) GetUserChatBoosts(ctx context.Context, chatID any, userID int64) (*types.UserChatBoosts, error) {
	payload := map[string]any{
		"chat_id": chatID,
		"user_id": userID,
	}
	var boosts types.UserChatBoosts
	if err := b.Request(ctx, "getUserChatBoosts", payload, &boosts); err != nil {
		return nil, err
	}
	return &boosts, nil
}

File: packages/python/src/telebot_py/bot/chat_management.py
"""Chat management Bot API methods (parity with packages/go/pkg/bot/chat_management.go)."""

from __future__ import annotations

import typing as t

from telebot_py.bot.base import (
    MarkupLike,
    Requester,
    clean_payload,
    parse_flag,
    parse_result,
    parse_string,
    to_wire,
)
from telebot_py.types.base import TypeParseError
from telebot_py.types.chat import ChatAdministratorRights
from telebot_py.types.topics import (
    MenuButton,
    MenuButtonCommands,
    MenuButtonDefault,
    MenuButtonWebApp,
)


class ChatManagementMixin(Requester):
    """Bot methods for chat titles, photos, pins, permissions, and settings."""

    async def set_chat_title(self, chat_id: int | str, title: str) -> bool:
        """Change the title of a chat.

        Titles can't be changed for private chats; the bot must be an
        administrator.

        Example:
            >>> ok = await bot.set_chat_title(-100, "New Title")

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target channel.
            title: New chat title; 0-128 characters.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id, title=title)
        return parse_flag(await self.request("setChatTitle", payload))

    async def set_chat_description(self, chat_id: int | str, description: str) -> bool:
        """Change the description of a group, supergroup, or channel.

        Example:
            >>> ok = await bot.set_chat_description("@channel", "New description")

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target channel.
            description: New chat description; 0-255 characters.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id, description=description)
        return parse_flag(await self.request("setChatDescription", payload))

    async def set_chat_photo(self, chat_id: int | str, photo: str) -> bool:
        """Set a new profile photo for the chat.

        Photos can't be changed for private chats; the bot must be an
        administrator. Accepts a ``file_id`` string; multipart uploads are
        intentionally out of scope (JSON payloads only).

        Example:
            >>> ok = await bot.set_chat_photo(-100, "photo_file_id")

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target channel.
            photo: New chat photo ``file_id``.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id, photo=photo)
        return parse_flag(await self.request("setChatPhoto", payload))

    async def delete_chat_photo(self, chat_id: int | str) -> bool:
        """Delete a chat photo.

        Photos can't be deleted for private chats; the bot must be an
        administrator.

        Example:
            >>> ok = await bot.delete_chat_photo(-100)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target channel.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id)
        return parse_flag(await self.request("deleteChatPhoto", payload))

    async def pin_chat_message(
        self,
        chat_id: int | str,
        message_id: int,
        *,
        disable_notification: bool | None = None,
    ) -> bool:
        """Add a message to the list of pinned messages in a chat.

        Example:
            >>> ok = await bot.pin_chat_message(-100, 42)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target channel.
            message_id: Identifier of the message to pin.
            disable_notification: Pin without sending a notification.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(
            chat_id=chat_id,
            message_id=message_id,
            disable_notification=disable_notification,
        )
        return parse_flag(await self.request("pinChatMessage", payload))

    async def unpin_chat_message(self, chat_id: int | str, message_id: int | None = None) -> bool:
        """Remove a message from the list of pinned messages in a chat.

        Example:
            >>> ok = await bot.unpin_chat_message(-100, 42)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target channel.
            message_id: Identifier of the message to unpin; omit to unpin the
                most recently pinned message.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id, message_id=message_id)
        return parse_flag(await self.request("unpinChatMessage", payload))

    async def unpin_all_chat_messages(self, chat_id: int | str) -> bool:
        """Clear the list of pinned messages in a chat.

        Example:
            >>> ok = await bot.unpin_all_chat_messages(-100)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target channel.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id)
        return parse_flag(await self.request("unpinAllChatMessages", payload))

    async def set_chat_permissions(
        self,
        chat_id: int | str,
        permissions: MarkupLike,
        *,
        use_independent_chat_permissions: bool | None = None,
    ) -> bool:
        """Set default chat permissions for all members.

        The bot must be an administrator in the group or a supergroup.

        Example:
            >>> ok = await bot.set_chat_permissions(-100, {"can_send_messages": True})

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.
            permissions: New default permissions as a ChatPermissions dict or
                ``to_dict`` object.
            use_independent_chat_permissions: Pass True if permissions apply
                independently of chat-level permissions.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(
            chat_id=chat_id,
            permissions=to_wire(permissions),
            use_independent_chat_permissions=use_independent_chat_permissions,
        )
        return parse_flag(await self.request("setChatPermissions", payload))

    async def export_chat_invite_link(self, chat_id: int | str) -> str:
        """Generate a new primary invite link for a chat.

        Any previously generated primary link is revoked. The bot must be an
        administrator in the chat.

        Example:
            >>> link = await bot.export_chat_invite_link(-100)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target channel.

        Returns:
            The exported invite link.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id)
        return parse_string(await self.request("exportChatInviteLink", payload))

    async def set_chat_menu_button(
        self, chat_id: int | None = None, menu_button: MarkupLike | None = None
    ) -> bool:
        """Change the bot's menu button in a private chat, or the default one.

        Example:
            >>> ok = await bot.set_chat_menu_button(42, {"type": "commands"})

        Args:
            chat_id: Unique identifier of the target private chat; omit for
                the default menu button.
            menu_button: New menu button as a MenuButton dict or ``to_dict``
                object.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id, menu_button=to_wire(menu_button))
        return parse_flag(await self.request("setChatMenuButton", payload))

    async def get_chat_menu_button(self, chat_id: int | None = None) -> MenuButton:
        """Get the current value of the bot's menu button.

        Example:
            >>> button = await bot.get_chat_menu_button()

        Args:
            chat_id: Unique identifier of the target private chat; omit for
                the default menu button.

        Returns:
            The MenuButton variant currently set.

        Raises:
            TypeParseError: If Telegram returns an unsupported menu button type.
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id)
        result = await self.request("getChatMenuButton", payload)
        if not isinstance(result, t.Mapping):
            msg = f"expected JSON object for MenuButton, got {type(result).__name__}"
            raise TypeParseError(msg)
        data = t.cast("t.Mapping[str, object]", result)
        button_type = data.get("type")
        if button_type == "commands":
            return MenuButtonCommands.from_dict(data)
        if button_type == "web_app":
            return MenuButtonWebApp.from_dict(data)
        if button_type == "default":
            return MenuButtonDefault.from_dict(data)
        msg = f"unsupported menu button type: {button_type}"
        raise TypeParseError(msg)

    async def set_my_default_administrator_rights(
        self,
        rights: MarkupLike,
        *,
        for_channels: bool | None = None,
    ) -> bool:
        """Change the default administrator rights of the bot.

        Example:
            >>> ok = await bot.set_my_default_administrator_rights(rights_dict)

        Args:
            rights: New default rights as a ChatAdministratorRights dict or
                ``to_dict`` object.
            for_channels: Pass True to change rights for channel posts.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(rights=to_wire(rights), for_channels=for_channels)
        return parse_flag(await self.request("setMyDefaultAdministratorRights", payload))

    async def get_my_default_administrator_rights(
        self, *, for_channels: bool | None = None
    ) -> ChatAdministratorRights:
        """Get the current default administrator rights of the bot.

        Example:
            >>> rights = await bot.get_my_default_administrator_rights()

        Args:
            for_channels: Pass True to get rights for channel posts.

        Returns:
            The ChatAdministratorRights currently set as the default.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(for_channels=for_channels)
        return parse_result(
            ChatAdministratorRights,
            await self.request("getMyDefaultAdministratorRights", payload),
        )

    async def set_chat_sticker_set(self, chat_id: int | str, sticker_set_name: str) -> bool:
        """Set a new group sticker set for a supergroup.

        The bot must be an administrator in the chat for this to work and must
        have the ``can_change_info`` administrator right. Use the field
        ``active_sticker_set_id`` returned by ``get_chat`` to test whether the
        bot can set the sticker set.

        Example:
            >>> ok = await bot.set_chat_sticker_set(-100, "test_set_by_bot")

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.
            sticker_set_name: Name of the sticker set to be set as the group
                sticker set.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id, sticker_set_name=sticker_set_name)
        return parse_flag(await self.request("setChatStickerSet", payload))

    async def delete_chat_sticker_set(self, chat_id: int | str) -> bool:
        """Delete the group sticker set from a supergroup.

        The bot must be an administrator in the chat for this to work and must
        have the appropriate administrator rights.

        Example:
            >>> ok = await bot.delete_chat_sticker_set(-100)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id)
        return parse_flag(await self.request("deleteChatStickerSet", payload))
